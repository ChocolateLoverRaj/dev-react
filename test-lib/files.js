export class File {
  constructor () {
    this.canRead = true
    this.canWrite = true
  }
}

export class NormalFile extends File {
  constructor (content) {
    super()
    this.content = content
  }
}

export class Dir extends File {
  constructor () {
    super()
    this.files = new Map()
  }
}

class EnoentError extends Error {
  constructor () {
    super(...arguments)
    this.code = 'ENOENT'
  }
}

export const topDir = new Dir()

export const reset = () => {
  topDir.files.clear()
}

export const getFile = path => {
  let file = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (file.files.has(filename)) {
      file = file.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
  return file
}

export const setFile = (path, fileToSet) => {
  let dir = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (fileTree.length === filenames.length) {
      dir.files.set(filename, fileToSet)
      return
    }
    if (dir.files.has(filename)) {
      dir = dir.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
}

export const unlinkFile = path => {
  let file = topDir
  const filenames = path.split('/')
  let fileTree = []
  for (const filename of filenames) {
    fileTree.push(filename)
    if (file.files.has(filename)) {
      if (fileTree.length === filenames.length) {
        file.files.delete(filename)
        return
      }
      file = file.files.get(filename)
    } else {
      throw new EnoentError(`File: ${fileTree.join('/')} does not exist`)
    }
  }
}
