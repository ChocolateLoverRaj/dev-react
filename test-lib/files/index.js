// TODO: structure with folders and files
const topDir = new Map()

export default topDir