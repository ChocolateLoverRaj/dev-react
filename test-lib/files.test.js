/* global afterEach, describe, test, expect */

import { File, NormalFile, Dir, topDir, getFile, reset, setFile, unlinkFile } from './files.js'

afterEach(reset)

describe('inheritance', () => {
  test('File', () => {
    const file = new File()
    expect(file).toBeInstanceOf(File)
    expect(file).not.toBeInstanceOf(NormalFile)
    expect(file).not.toBeInstanceOf(Dir)
  })

  test('NormalFile', () => {
    const file = new NormalFile('content')
    expect(file).toBeInstanceOf(File)
    expect(file).toBeInstanceOf(NormalFile)
    expect(file).not.toBeInstanceOf(Dir)
  })

  test('Dir', () => {
    const file = new Dir()
    expect(file).toBeInstanceOf(File)
    expect(file).not.toBeInstanceOf(NormalFile)
    expect(file).toBeInstanceOf(Dir)
  })
})

const testFiles = createFile => {
  function topLevel () {
    const file = createFile()
    topDir.files.set('file', file)
    expect(getFile('file')).toBe(file)
  }
  function subDir () {
    const file = createFile()
    const dir = new Dir()
    dir.files.set('file', file)
    topDir.files.set('dir', dir)
    expect(getFile('dir/file')).toBe(file)
  }
  function subSubDir () {
    const file = createFile()
    const subDir = new Dir()
    subDir.files.set('file', file)
    const dir = new Dir()
    dir.files.set('subDir', subDir)
    topDir.files.set('dir', dir)
    expect(getFile('dir/subDir/file')).toBe(file)
  }
  topLevel()
  subDir()
  subSubDir()
}

describe('getFile', () => {
  describe('nonexistent file', () => {
    test('in topDir', () => {
      expect(() => {
        getFile('file')
      }).toThrowErrorMatchingSnapshot()
    })

    test('in subDir', () => {
      topDir.files.set('dir', new Dir())
      expect(() => {
        getFile('dir/file')
      }).toThrowErrorMatchingSnapshot()
    })

    test('in nonexistent subDir', () => {
      expect(() => {
        getFile('dir/file')
      }).toThrowErrorMatchingSnapshot()
    })
  })

  test('get NormalFile', () => {
    testFiles(() => new NormalFile('hi'))
  })

  test('get Dir', () => {
    testFiles(() => new Dir())
  })
})

describe('setFile', () => {
  test('top level', () => {
    setFile('file', new NormalFile('hi'))
    expect(topDir.files.get('file').content).toBe('hi')
  })

  test('dir level', () => {
    setFile('dir', new Dir())
    setFile('dir/file', new NormalFile('hi'))
    expect(topDir.files.get('dir').files.get('file').content).toBe('hi')
  })

  test('overwrite', () => {
    setFile('file', new NormalFile('hi'))
    expect(topDir.files.get('file').content).toBe('hi')
    setFile('file', new NormalFile('hello'))
    expect(topDir.files.get('file').content).toBe('hello')
  })
})

test('reset', () => {
  const topLevelFile = new NormalFile('hi')
  const dirFile = new NormalFile('hello')
  setFile('file', topLevelFile)
  setFile('dir', new Dir())
  setFile('dir/file', dirFile)
  reset()
  expect(topDir.files.size).toBe(0)
})

describe('unlink', () => {
  test('top dir', () => {
    const file = new NormalFile('hi')
    setFile('file', file)
    unlinkFile('file')
    expect(topDir.files.has('file')).toBe(false)
  })

  test('sub dir', () => {
    const file = new NormalFile('hi')
    setFile('dir', new Dir())
    setFile('dir/file', file)
    unlinkFile('dir/file')
    expect(topDir.files.get('dir').files.has('file')).toBe(false)
  })
})
