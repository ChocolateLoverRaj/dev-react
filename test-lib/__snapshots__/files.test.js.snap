// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFile nonexistent file in nonexistent subDir 1`] = `"File: dir does not exist"`;

exports[`getFile nonexistent file in subDir 1`] = `"File: dir/file does not exist"`;

exports[`getFile nonexistent file in topDir 1`] = `"File: file does not exist"`;
